sparse/coo.py

```diff
@@ -2511,8 +2512,39 @@ def _linear_loc(coords, shape, signed=False):
     tmp = np.zeros(coords.shape[1], dtype=dtype)
     strides = 1
     for i, d in enumerate(shape[::-1]):
-        # out += self.coords[-(i + 1), :].astype(dtype) * strides
         np.multiply(coords[-(i + 1), :], strides, out=tmp, dtype=dtype)
         np.add(tmp, out, out=out)
         strides *= d
     return out
+
+
+def _posarg_partial(func, pos, posargs):
+    if not isinstance(pos, Iterable):
+        pos = (pos,)
+        posargs = (posargs,)
+
+    n_partial_args = len(pos)
+
+    class Partial(object):
+        def __call__(self, *args, **kwargs):
+            j = 0
+            totargs = []
+
+            for i in range(len(args) + n_partial_args):
+                if j >= n_partial_args or i != pos[j]:
+                    totargs.append(args[i - j])
+                else:
+                    totargs.append(posargs[j])
+                    j += 1
+
+            return func(*totargs, **kwargs)
+
+        def __str__(self):
+            return str(func)
+
+        def __repr__(self):
+            return repr(func)
+
+        __doc__ = func.__doc__
+
+    return Partial()
```

Comment:
I could have, but our situation is slightly unique:

1. We're using `str(func)` for exceptions. `functools.wraps` doesn't work on that for all callables (e.g. `ufunc`s), and breaks a few docstrings. This leads to illegible names in exceptions like `_posarg_partial.<locals>.wrapper` (and the same for debugging).
2. We're replacing a number of arguments in different positions.

I guess I could turn it into a class rather than a decorator style function.

Reply:
I turned it into a callable class.